### DISCUSSION TITLE
0.8 Release Plan

### DISCUSSION THREAD

**Discussion by brichet** [jupyter/nbgrader#1634] [General]
This is an issue for 0.8 release plan, to port nbgrader extensions to notebook 7

Classic notebook extensions system and notebook 7 extensions system (labextension) are not compatibles.
Consequently, the first step would be to move the current master branch to a 0.7.x branch (a branch which keep compatibility with Classic Notebook), and use the master branch  for 0.8 release (compatible with Notebook 7 and Jupyterlab).

That 0.8 release would have the following steps :

- [x] move current `master` branch to `0.7.x` branch
- [x] rename `master` branch to `main`
- [x] adopt juputer releaser (#1570)
- [x] migrate server extension to jupyter_server (#1465)
- [x] port Notebook extensions to Jupyterlab extensions (#1006, #1550)

**Comment by jtpio** [comment]
Thanks @brichet for starting this!

> move current master branch to 0.6.x branch

It would also be great to take the opportunity to rename the `master` branch to `main`. So it's consistent with most of the other Jupyter repos.

**Comment by jtpio** [comment]
It looks like there is already a `0.7.0` milestone, with a couple of PRs merged in `master`: https://github.com/jupyter/nbgrader/milestone/23?closed=1

@BertR or anyone else recently active, do you know if there was a plan to make a `0.7.0` release with these changes?

If so, maybe we'll want to make the release before branching and continuing further work as mentioned in the comment above?

**Comment by jtpio** [comment]
> @BertR or anyone else recently active, do you know if there was a plan to make a `0.7.0` release with these changes?

cc @jhamrick who might also know about a previous 0.7.0 release plan.

**Comment by brichet** [comment]
Thanks @jtpio, I had not seen that `0.7.0` milestone.
If this is still planned, fixing nbconvert dependency could be there as it is an interesting feature for both notebook classic and jupyterlab/notebook 7.

**Comment by jhamrick** [comment]
IIRC, the main plan for 0.7.0 was to fix the breakage with nbconvert (which I started with #1421 but was unable to complete). I think doing effectively a maintenance release for 0.7.0 that fixes that breakage (and any others that might have popped up) would be fantastic, and then do a subsequent release that does the update for JupyterLab extensions could make sense---then there's one final release with up-to-date dependencies without JupyterLab, and a second release that does the migration to JupyterLab.

Thanks so much to everyone working on this, it's fantastic to see some progress towards this 😄

**Comment by jhamrick** [comment]
Actually looking into things a bit more, I am now reminded that there's been a lot of work for 0.7.0 and #1421 was basically the last blocking thing---so I would definitely recommend doing that fix, then releasing 0.7.0, then starting the new functionality for 0.8.0.

I'll update the labels for various PRs to make it clearer that everything merged so far should be for the 0.7.0 release...

**Comment by brichet** [comment]
@jhamrick #1567 was merged yesterday to fix the breakage with nbconvert. It should have start from #1421 but it seems that I didn't seen how advanced it was.
I guess this means we can release 0.7.0 and start 0.8.0 release for jupyterlab extensions.

**Comment by brichet** [comment]
I renamed the issue to **0.8 release plan** as the 0.7 already exists.

**Comment by jhamrick** [comment]
Fantastic, thanks! Yeah I initially missed #1567 too but really glad you were able to get it fixed. Exciting to see things moving forward 😄

**Comment by jhamrick** [comment]
I am not sure, but perhaps it is worth trying to resolve https://github.com/jupyter/nbgrader/pull/1557 for 0.7 as well? Just wanted to flag it as I noticed it caused the tests to break as well.

**Comment by brichet** [comment]
The tests pass on my side with python 3.8 and 3.9, and `jupyter-client==7.3.0`.
Is it possible to redo these tests with the latest master version or do we need to open a PR ?

**Comment by jhamrick** [comment]
Ah that's great. We should be able to do it with dependabot commands.

**Comment by jtpio** [comment]
So as a summary, how about the following?

- merge https://github.com/jupyter/nbgrader/pull/1557 if ready
- check if other PRs should go in `0.7.0`
- release 0.7.0 from `master`, or maybe a pre-release first
- rename master to `main`
- create the `0.7.x` branch to continue making releases for the 0.7.x series
- main targets `0.8.0` and we continue there for the JupyterLab related work

Maybe 0.7.0 can be released manually as before. And we switch to the releaser (https://github.com/jupyter/nbgrader/issues/1570) for 0.8.0?

**Comment by jhamrick** [comment]
Sounds like a good plan to me!

**Comment by jhamrick** [comment]
I created #1573 to prepare the 0.7.0 release. I will try to actually make the release tomorrow, though I am not 100% sure I'll have the time. If not, feel free to go ahead and do the release in my absence @jtpio (basically after this PR is merged following all steps from https://nbgrader.readthedocs.io/en/stable/contributor_guide/releasing.html#clean-all-untracked-files onwards).

**Comment by jhamrick** [comment]
0.7.0 is released! I still have to do the conda package but otherwise you should be good to proceed with all the stuff for 0.8.0 😄

**Comment by jtpio** [comment]
Nice, thanks @jhamrick!

**Comment by brichet** [comment]
@jhamrick @jtpio do you think there are missing features for a pre-release ?

**Comment by jtpio** [comment]
A first `0.8.0a0` pre-release sounds good, so it can more easily be tested outside of the repo :+1:

**Comment by jhamrick** [comment]
Sure, sounds good to me. Perhaps it's also worth doing a release for 0.7.1 now too... I can do the 0.7.1 release and feel free to go ahead with the 0.8.0a0 pre-release.

**Comment by jtpio** [comment]
> go ahead with the 0.8.0a0 pre-release.

Looking into it with the Jupyter Releaser.

**Comment by jtpio** [comment]
The first `0.8.0a0` is out! https://github.com/jupyter/nbgrader/releases/tag/v0.8.0a0

Try it with `pip install --pre nbgrader`.

**Comment by brichet** [comment]
Thanks @jtpio.
And the next one should be easier now !

**Comment by brichet** [comment]
It seems that the labextension published with the package kept the version `0.8.0.dev0`, according to that error message (which also have to be adapted to labextension...  https://github.com/jupyter/nbgrader/pull/1616) :
`The version of the Assignment List nbextension does not match the server extension; the nbextension version is 0.8.0.dev0 while the server version is 0.8.0a0`

**Comment by jtpio** [comment]
Ah thanks @brichet for catching this.

For reference here is the log of the Full Release run: https://github.com/jtpio/jupyter_releaser/runs/7090980402?check_suite_focus=true

A guess would be that it's because the extension was built during the setup step, which produced the static assets for the prebuilt extension. And later the extension build would be skipped because of the `skip-if-exists` option from `jupyter-packaging`.

For notebook we use this releaser hook to make sure to start from a clean state before building the Python wheel and sdist, which is probably related iirc:

https://github.com/jupyter/notebook/blob/b9bab689c9a2f33eb3b2cca1383c2d99baa7a2e8/package.json#L70-L72

**Comment by brichet** [comment]
There is already something similar in nbgrader which clean the `lib` and the `nbgrader/labextension` directories :
https://github.com/jupyter/nbgrader/blob/09e632be783c88c9b1ee2f7fc46ed5130a596a51/package.json#L104-L106

**Comment by jtpio** [comment]
Ah ok then maybe it's just because this hook defined in `pyproject.toml` was not triggered?

https://github.com/jupyter/nbgrader/blob/09e632be783c88c9b1ee2f7fc46ed5130a596a51/pyproject.toml#L19-L20

If there is already some releaser configuration in `package.json`.

We should keep the configuration in one place only, for example `pyproject.toml`.

**Comment by brichet** [comment]
You're right, it's necessarily better to keep all the hooks in only one place.
I just opened https://github.com/jupyter/nbgrader/pull/1617.

**Comment by brichet** [comment]
Pre-release [0.8.0a2](https://github.com/jupyter/nbgrader/releases/tag/v0.8.0a2) is out, `pip install --pre nbgrader` to try it.

**Comment by brichet** [comment]
The release can also be tested on https://mybinder.org/v2/gh/jupyter/nbgrader.git/main
Thanks @jtpio for help.

**Comment by brichet** [comment]
Version `v0.8.0` is released : https://github.com/jupyter/nbgrader/releases/tag/v0.8.0

Installation : `pip install nbgrader`

Binder : https://mybinder.org/v2/gh/jupyter/nbgrader.git/v0.8.0

**Comment by jtpio** [comment]
Really nice, thanks @brichet and everyone who contributed!

**Comment by krassowski** [comment]
Is there going to be a blog post or other announcement about this release to share?

**Reply by brichet** [reply]
Yes, we plan to do a blog post about it.

**Reply by AstroMike** [reply]
That would be [here]( https://blog.jupyter.org/upgrading-nbgrader-99c56ae56c47), presumably. (Adding this mostly for those who find their way here via Google).
